File: apps/api/app/services/publishers/youtube.py
"""YouTube publisher — Shorts upload via resumable upload API."""

import logging
from datetime import datetime, timezone

import httpx

from app.services.publishers.base import BasePublisher, PublishResult
from app.utils.exceptions import ValidationError

logger = logging.getLogger(__name__)

UPLOAD_API = "https://www.googleapis.com/upload/youtube/v3/videos"
YOUTUBE_TITLE_MAX = 100
YOUTUBE_DESCRIPTION_MAX = 5000


class YouTubePublisher(BasePublisher):
    platform_id = "youtube"

    def validate(self, output: dict) -> None:
        media_urls = output.get("media_urls", [])
        if not media_urls:
            raise ValidationError(
                message="Video required",
                detail="YouTube Shorts requires a video file.",
            )
        title = output.get("title", "")
        if len(title) > YOUTUBE_TITLE_MAX:
            raise ValidationError(
                message="Title too long",
                detail=f"Title is {len(title)} chars (max {YOUTUBE_TITLE_MAX}).",
            )

    async def publish(self, output: dict, token: str, **kwargs) -> PublishResult:
        title = output.get("title", "Untitled Short")
        description = output.get("content", "")[:YOUTUBE_DESCRIPTION_MAX]
        video_url = output.get("media_urls", [""])[0]
        tags = output.get("tags", [])

        # Step 1: Download video from storage URL
        async with httpx.AsyncClient(timeout=60) as client:
            video_resp = await client.get(video_url)
        if video_resp.status_code != 200:
            return PublishResult(
                success=False, platform="youtube",
                error="Failed to download video from storage.",
            )
        video_bytes = video_resp.content

        # Step 2: Initiate resumable upload
        metadata = {
            "snippet": {
                "title": title,
                "description": description,
                "tags": tags,
                "categoryId": "22",  # People & Blogs
            },
            "status": {
                "privacyStatus": "public",
                "selfDeclaredMadeForKids": False,
                "madeForKids": False,
            },
        }

        async with httpx.AsyncClient(timeout=30) as client:
            init_resp = await client.post(
                UPLOAD_API,
                params={"uploadType": "resumable", "part": "snippet,status"},
                json=metadata,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json; charset=UTF-8",
                    "X-Upload-Content-Type": "video/*",
                    "X-Upload-Content-Length": str(len(video_bytes)),
                },
            )

        if init_resp.status_code != 200:
            logger.error("YouTube upload init failed: %s %s", init_resp.status_code, init_resp.text)
            return PublishResult(
                success=False, platform="youtube",
                error=f"Upload init failed: {init_resp.status_code}",
            )

        upload_url = init_resp.headers.get("location", "")
        if not upload_url:
            return PublishResult(
                success=False, platform="youtube",
                error="No upload URL returned from YouTube.",
            )

        # Step 3: Upload video bytes
        async with httpx.AsyncClient(timeout=300) as client:
            upload_resp = await client.put(
                upload_url,
                content=video_bytes,
                headers={"Content-Type": "video/*"},
            )

        if upload_resp.status_code not in (200, 201):
            logger.error("YouTube upload failed: %s %s", upload_resp.status_code, upload_resp.text)
            return PublishResult(
                success=False, platform="youtube",
                error=f"Upload failed: {upload_resp.status_code}",
            )

        data = upload_resp.json()
        video_id = data.get("id", "")
        return PublishResult(
            success=True,
            platform="youtube",
            platform_post_id=video_id,
            platform_post_url=f"https://youtube.com/shorts/{video_id}",
            published_at=datetime.now(timezone.utc),
            metadata={"format": "shorts"},
        )

File: functions/src/lib/platforms/publishers/youtube.ts
/**
 * YouTube Data API v3 publisher — Shorts and video upload.
 * Uses resumable upload protocol for reliability.
 */

import type { PlatformPublisher, PublishResult, DecryptedTokens, RefreshedTokens } from "./base.js";

const UPLOAD_URL = "https://www.googleapis.com/upload/youtube/v3/videos";
const TOKEN_URL = "https://oauth2.googleapis.com/token";

export class YouTubePublisher implements PlatformPublisher {
  async publish(content: string, metadata: Record<string, unknown>, tokens: DecryptedTokens): Promise<PublishResult> {
    if (!tokens.accessToken) {
      return { success: false, postId: null, url: null, error: "No YouTube access token. Reconnect your account." };
    }

    try {
      const title = (metadata.title as string) || content.substring(0, 100);
      const description = (metadata.description as string) || content;
      const tags = (metadata.tags as string[]) || [];
      const videoUrl = (metadata.video_url as string) || "";

      if (!videoUrl) {
        return {
          success: false,
          postId: null,
          url: null,
          error: "YouTube requires a video_url in metadata to upload.",
        };
      }

      // Download the video content
      const videoResp = await fetch(videoUrl);
      if (!videoResp.ok) {
        throw new Error(`Failed to fetch video from ${videoUrl}: ${videoResp.status}`);
      }
      const videoBuffer = await videoResp.arrayBuffer();

      // Step 1: Initialize resumable upload
      const videoMetadata = {
        snippet: {
          title: title.substring(0, 100),
          description: description.substring(0, 5000),
          tags: tags.slice(0, 30),
          categoryId: "22", // People & Blogs
        },
        status: {
          privacyStatus: "public",
          selfDeclaredMadeForKids: false,
          ...(metadata.is_short ? { shorts: { isShort: true } } : {}),
        },
      };

      const initResp = await fetch(
        `${UPLOAD_URL}?uploadType=resumable&part=snippet,status`,
        {
          method: "POST",
          headers: {
            Authorization: `Bearer ${tokens.accessToken}`,
            "Content-Type": "application/json; charset=UTF-8",
            "X-Upload-Content-Type": "video/*",
            "X-Upload-Content-Length": String(videoBuffer.byteLength),
          },
          body: JSON.stringify(videoMetadata),
        },
      );

      if (!initResp.ok) {
        const body = await initResp.text();
        throw new Error(`YouTube upload init failed ${initResp.status}: ${body}`);
      }

      const uploadUrl = initResp.headers.get("Location");
      if (!uploadUrl) {
        throw new Error("YouTube did not return a resumable upload URL.");
      }

      // Step 2: Upload the video data
      const uploadResp = await fetch(uploadUrl, {
        method: "PUT",
        headers: {
          "Content-Type": "video/*",
          "Content-Length": String(videoBuffer.byteLength),
        },
        body: videoBuffer,
      });

      if (!uploadResp.ok) {
        const body = await uploadResp.text();
        throw new Error(`YouTube upload failed ${uploadResp.status}: ${body}`);
      }

      const uploadData = await uploadResp.json() as Record<string, unknown>;
      const videoId = uploadData.id as string;

      return {
        success: true,
        postId: videoId,
        url: `https://www.youtube.com/watch?v=${videoId}`,
        error: null,
      };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return { success: false, postId: null, url: null, error: `YouTube publish error: ${message}` };
    }
  }

  async validateConnection(_tokens: DecryptedTokens, expiresAt: Date | null): Promise<boolean> {
    if (expiresAt) return expiresAt > new Date();
    return true;
  }

  async refreshToken(tokens: DecryptedTokens): Promise<RefreshedTokens | null> {
    if (!tokens.refreshToken) return null;

    const clientId = process.env.YOUTUBE_CLIENT_ID || "";
    const clientSecret = process.env.YOUTUBE_CLIENT_SECRET || "";

    if (!clientId || !clientSecret) return null;

    const resp = await fetch(TOKEN_URL, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: tokens.refreshToken,
        client_id: clientId,
        client_secret: clientSecret,
      }),
    });

    if (!resp.ok) return null;

    const data = await resp.json() as Record<string, unknown>;
    return {
      accessToken: data.access_token as string,
      refreshToken: (data.refresh_token as string) ?? tokens.refreshToken,
      expiresIn: (data.expires_in as number) ?? 3600,
    };
  }
}
